stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -745,8 +742,10 @@ extension ${Self}: BinaryFloatingPoint {
     if isZero { return .min }
     let provisional = Int(exponentBitPattern) - Int(${Self}._exponentBias)
     if isNormal { return provisional }
-    let shift = ${Self}.significandBitCount - significandBitPattern.signBitIndex
-    return provisional + 1 - Int(shift)
+    let shift =
+      ${Self}.significandBitCount &-
+        (${RawSignificand}.bitWidth &- significandBitPattern.leadingZeroBitCount)
```

Comment:
In the other locations, you changed `signBitIndex` to `_binaryLogarithm()`; why is it handled differently here?

Reply:
We get to avoid the dance of subtracting 1, casting to `RawSignificand`, casting back to `Int`, then adding 1. And in the process, the mechanics of what we're doing here become less opaque to the reader and therefore easier to reason.

Reply:
Looking at this more, this is a pretty big clarity loss in this case. `shift` is so named because it is the number of bits that the significand needs to be left-shifted to put the leading bit in the position of the implicit bit of a normal number. With your change, `shift` no longer represents that value; it's off by one.

Similarly, the extra `+1` directly corresponds to the fact that subnormal numbers have an exponent field that's one off from the exponent of the least normal binade.

Reply:
Yet it's not arbitrary or a coincidence that shifting by one less and not adding one to the exponent cancel each other out here.

The fact that incrementing the bit pattern of the greatest subnormal magnitude and least normal magnitude are, respectively,

```swift
0b0_00000000000_1111111111111111111111111111111111111111111111111111
0b0_00000000001_0000000000000000000000000000000000000000000000000000
```
(in other words, the fact that they differ by 1) _requires_ both of the properties you enumerate to cancel each other out in terms of computation of the exponent. Therefore, it is perfectly correct to consider the adjustment required to yield the correct exponent to be adding to `provision` that value by which the significand must be left-shifted such that the explicit leading bit is in most significant position of the significand bit pattern.